lti_consumer/migrations/0014_add_remaining_fields_to_lti_config.py

```diff
@@ -0,0 +1,33 @@
+# Generated by Django 3.2.13 on 2022-05-06 19:31
+
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('lti_consumer', '0013_auto_20210712_1352'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='lticonfiguration',
+            name='lti_1p3_launch_url',
+            field=models.CharField(blank=True, help_text='This is the LTI launch URL, otherwise known as the target_link_uri.It represents the LTI resource to launch to in the second leg of the launch flow, when the resource is actually launched.', max_length=255),
+        ),
+        migrations.AddField(
+            model_name='lticonfiguration',
+            name='lti_1p3_oidc_url',
+            field=models.CharField(blank=True, help_text='This is the OIDC login intitiation URL in the LTI 1.3 flow, which should be provided by the LTI Tool.', max_length=255),
+        ),
+        migrations.AddField(
+            model_name='lticonfiguration',
+            name='lti_1p3_tool_keyset_url',
+            field=models.CharField(blank=True, help_text="This is the Tool's JWK (JSON Web Key) keyset URL. This should be provided by the LTI Tool. One of either lti_1p3_tool_public_key or lti_1p3_tool_keyset_url must not be blank.", max_length=255),
```

Comment:
The `model.clean()` method is one form of model validation. It's called as part of a call to `model.full_clean()`, which calls `model.clean_fields()`, `model.clean()`, and `model.validate_unique()`. You can call `model.clean()` on its own, but it will also be invoked when calling `model.full_clean()`.

When calling `is_valid()` on a `ModelForm`, `model.full_clean()` will be called. Because the Django admin form uses ModelForms, `clean` will be called when saving an object in the Django admin, which is what we want for new objects.

As far as existing objects, if someone goes into the admin and tries to modify the existing object, then they will see an error in the Django admin.

That said, I don't believe that it's even possible to use LTI consumers with the config_store `CONFIG_ON_DB` right now, so this should not occur.

There are 19 objects in the database with the `config_store` set to `CONFIG_ON_DB`, so once we use `CONFIG_ON_DB`, we should determine what they're used for, if at all. If necessary, we can remove them or can manually update these values.

As an aside, `model.clean()` is not called when calling `save()` on a model.

Reply:
>That said, I don't believe that it's even possible to use LTI consumers with the config_store CONFIG_ON_DB right now, so this should not occur.

LTI 1.3 doesn't support `CONFIG_ON_DB` now but that's going to change in the next few weeks - I and @tecoholic  (OpenCraft) are working on pluggable LTI configurations for LTI 1.3 in the current sprint.

`CONFIG_ON_DB` can **currently** be used for any kind of LTI 1.1 integration or embedded anywhere in the platform, and if I'm not mistaken it's used by the new Discussions project integrations. Take care to only delete the `LTI_1P3` versions with `CONFIG_ON_DB` to avoid potentially breaking other integrations.

Reply:
Got it. In that case, we should probably just leave them alone and not delete anything. I modified the condition to only raise an exception if the `version` is 1.3, not 1.1. Configurations using LTI 1.1 will be unaffected. If someone tries to change a configuration using LTI 1.3, they'll just have to supply one of these values or reach out to support.